src/Functions/LowerUpperUTF8Impl.h

```diff
@@ -102,179 +42,37 @@ struct LowerUpperUTF8Impl
+        res_offsets.resize_exact(offsets.size());
+        size_t curr_offset = 0;
+        for (size_t i = 0; i < offsets.size(); ++i)
         {
-            if (*src >= not_case_lower_bound && *src <= not_case_upper_bound)
-                *dst++ = *src++ ^ flip_case_mask;
+            const auto * pos =  reinterpret_cast<const char *>(&data[offsets[i-1]]);
+            size_t size = offsets[i] - offsets[i-1];
+
+            icu::UnicodeString input(pos, static_cast<int32_t>(size), "UTF-8");
+            if constexpr (upper)
+                input.toUpper();
             else
-                *dst++ = *src++;
-        }
-        else if (src + 1 < src_end
-            && ((src[0] == 0xD0u && (src[1] >= 0x80u && src[1] <= 0xBFu)) || (src[0] == 0xD1u && (src[1] >= 0x80u && src[1] <= 0x9Fu))))
-        {
-            cyrillic_to_case(src, dst);
-        }
-        else if (src + 1 < src_end && src[0] == 0xC2u)
-        {
-            /// Punctuation U+0080 - U+00BF, UTF-8: C2 80 - C2 BF
-            *dst++ = *src++;
-            *dst++ = *src++;
-        }
-        else if (src + 2 < src_end && src[0] == 0xE2u)
-        {
-            /// Characters U+2000 - U+2FFF, UTF-8: E2 80 80 - E2 BF BF
-            *dst++ = *src++;
-            *dst++ = *src++;
-            *dst++ = *src++;
-        }
-        else
-        {
-            size_t src_sequence_length = UTF8::seqLength(*src);
-            /// In case partial buffer was passed (due to SSE optimization)
-            /// we cannot convert it with current src_end, but we may have more
-            /// bytes to convert and eventually got correct symbol.
-            if (partial && src_sequence_length > static_cast<size_t>(src_end - src))
-                return false;
+                input.toLower();
 
-            auto src_code_point = UTF8::convertUTF8ToCodePoint(src, src_end - src);
-            if (src_code_point)
-            {
-                int dst_code_point = to_case(*src_code_point);
-                if (dst_code_point > 0)
-                {
-                    size_t dst_sequence_length = UTF8::convertCodePointToUTF8(dst_code_point, dst, src_end - src);
-                    assert(dst_sequence_length <= 4);
+            String output;
+            input.toUTF8String(output);
+            const char * end = find_first_symbols<'\0'>(output.data(), output.data() + output.size());
+            size_t valid_size = end - output.data();
 
-                    /// We don't support cases when lowercase and uppercase characters occupy different number of bytes in UTF-8.
-                    /// As an example, this happens for ß and ẞ.
-                    if (dst_sequence_length == src_sequence_length)
-                    {
-                        src += dst_sequence_length;
-                        dst += dst_sequence_length;
-                        return true;
-                    }
-                }
-            }
+            res_data.resize(curr_offset + valid_size + 1);
```

Comment:
This potentially calls `.resize()` on the result structure once per loop iteration (l. 48). What we should do instead is `.reserve()` 2x the current space if the string-to-copy does not fit into the current capacity (i.e. exponential resizing). That way, resizing happens less frequently.

Reply:
resize has 2 steps 
1. reserve 2x current space if current space is not enough
2. execute `resize_assume_reserved` 

The second step is too trivial. The first step doesn't happen frequently. So I believe it it not necessary to `reserve` in advance because you never known what is the exact size of utf8 string after upper/lower conversion.

Reply:
Ah, I was not aware that `resize()` does an exponential resize internally. We are good then.